File: bootstrap/src/main/java/coursier/IsolatedClassLoader.java
package coursier;

import java.net.URL;
import java.net.URLClassLoader;

public class IsolatedClassLoader extends URLClassLoader {

    private String[] isolationTargets;

    public IsolatedClassLoader(
            URL[] urls,
            ClassLoader parent,
            String[] isolationTargets
    ) {
        super(urls, parent);
        this.isolationTargets = isolationTargets;
    }

    /**
     * Applications wanting to access an isolated `ClassLoader` should inspect the hierarchy of
     * loaders, and look into each of them for this method, by reflection. Then they should
     * call it (still by reflection), and look for an agreed in advance target in it. If it is found,
     * then the corresponding `ClassLoader` is the one with isolated dependencies.
     */
    public String[] getIsolationTargets() {
        return isolationTargets;
    }

}

File: cli/src/main/scala-2.12/coursier/cli/IsolatedClassLoader.scala
package coursier.cli

import java.net.{URL, URLClassLoader}


class IsolatedClassLoader(
  urls: Array[URL],
  parent: ClassLoader,
  isolationTargets: Array[String]
) extends URLClassLoader(urls, parent) {

  /**
    * Applications wanting to access an isolated `ClassLoader` should inspect the hierarchy of
    * loaders, and look into each of them for this method, by reflection. Then they should
    * call it (still by reflection), and look for an agreed in advance target in it. If it is found,
    * then the corresponding `ClassLoader` is the one with isolated dependencies.
    */
  def getIsolationTargets: Array[String] = isolationTargets

}